nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/TestGetOrCreateDocument.java

```diff
@@ -0,0 +1,229 @@
+
+import org.apache.commons.lang3.mutable.MutableObject;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.nuxeo.ecm.core.api.ConcurrentUpdateException;
+import org.nuxeo.ecm.core.api.CoreInstance;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.DocumentRef;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.kv.KeyValueService;
+import org.nuxeo.ecm.core.kv.KeyValueStore;
+import org.nuxeo.ecm.core.test.CoreFeature;
+import org.nuxeo.ecm.core.test.annotations.Granularity;
+import org.nuxeo.ecm.core.test.annotations.RepositoryConfig;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.test.runner.Deploy;
+import org.nuxeo.runtime.test.runner.Features;
+import org.nuxeo.runtime.test.runner.FeaturesRunner;
+import org.nuxeo.runtime.transaction.TransactionHelper;
+
+import javax.inject.Inject;
+
+import java.util.concurrent.CountDownLatch;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.nuxeo.ecm.core.api.CoreInstance.DOCUMENT_LOCK;
+
+/**
+ * @since 9.3
+ */
+@RunWith(FeaturesRunner.class)
+@Features(CoreFeature.class)
+@RepositoryConfig(cleanup = Granularity.METHOD)
+@Deploy("org.nuxeo.ecm.core.kv")
+public class TestGetOrCreateDocument {
+
+    @Inject
+    protected CoreFeature coreFeature;
+
+    @Inject
+    protected CoreSession session;
+
+    protected class GetOrCreateDocumentThread implements Runnable {
+
+        protected CountDownLatch latch;
+
+        protected DocumentModel documentModel;
+
+        protected CoreSession session;
+
+        public GetOrCreateDocumentThread(CoreSession session, DocumentModel documentModel) {
+            this(session, documentModel, null);
+        }
+
+        public GetOrCreateDocumentThread(CoreSession session, DocumentModel documentModel, CountDownLatch latch) {
+            this.session = session;
+            this.latch = latch;
+            this.documentModel = documentModel;
+        }
+
+        @Override
+        public void run() {
+
+            TransactionHelper.runInTransaction(() -> {
+                try {
+                    documentModel = CoreInstance.doAtomically(session, documentModel,
+                            TestGetOrCreateDocument.this::computeKey, (CoreSession s, DocumentModel docModel) -> {
+                                DocumentRef ref = docModel.getRef();
+                                if (s.exists(ref)) {
+                                    return s.getDocument(ref);
+                                }
+                                if (latch != null) {
+                                    // Simulate a thread that keeps a document locked
+                                    try {
+                                        latch.await();
```

Comment:
After thinking about it I think maybe a `CyclicBarrier` is a better wait primitive here, because I don't see how your last unit test `testExceptionWhenCreatingDocumentInAnotherThread` can work if the two threads don't wait for each other.